aoc2018/src/main/scala/day19.scala

```diff
@@ -0,0 +1,113 @@
+
+
+object day19 extends App:
+
+  private val day: String =
+    this.getClass.getName.drop(3).init
+
+  private val start1: Long =
+    System.currentTimeMillis
+
+  case class Operation(op: String, a: Int, b: Int, c: Int):
+    def addr(in: Array[Int]): Array[Int] = {in(c) = in(a) + in(b); in}
+    def addi(in: Array[Int]): Array[Int] = {in(c) = in(a) + b; in}
+    def mulr(in: Array[Int]): Array[Int] = {in(c) = in(a) * in(b); in}
+    def muli(in: Array[Int]): Array[Int] = {in(c) = in(a) * b; in}
+    def banr(in: Array[Int]): Array[Int] = {in(c) = in(a) & in(b); in}
+    def bani(in: Array[Int]): Array[Int] = {in(c) = in(a) & b; in}
+    def borr(in: Array[Int]): Array[Int] = {in(c) = in(a) | in(b); in}
+    def bori(in: Array[Int]): Array[Int] = {in(c) = in(a) | b; in}
+    def setr(in: Array[Int]): Array[Int] = {in(c) = in(a); in}
+    def seti(in: Array[Int]): Array[Int] = {in(c) = a; in}
+    def gtir(in: Array[Int]): Array[Int] = {in(c) = if a > in(b) then 1 else 0; in}
+    def gtri(in: Array[Int]): Array[Int] = {in(c) = if in(a) > b then 1 else 0; in}
+    def gtrr(in: Array[Int]): Array[Int] = {in(c) = if in(a) > in(b) then 1 else 0; in}
+    def eqir(in: Array[Int]): Array[Int] = {in(c) = if a == in(b) then 1 else 0; in}
+    def eqri(in: Array[Int]): Array[Int] = {in(c) = if in(a) == b then 1 else 0; in}
+    def eqrr(in: Array[Int]): Array[Int] = {in(c) = if in(a) == in(b) then 1 else 0; in}
+
+    val ops: Map[String, Array[Int] => Array[Int]] = Map(
+      "addr" -> addr, "addi" -> addi, "mulr" -> mulr, "muli" -> muli, "banr" -> banr, "bani" -> bani,
+      "borr" -> borr, "bori" -> bori, "setr" -> setr, "seti" -> seti, "gtir" -> gtir, "gtri" -> gtri,
+      "gtrr" -> gtrr, "eqir" -> eqir, "eqri" -> eqri, "eqrr" -> eqrr)
+
+    def run(input: Vector[Int]): Vector[Int] =
+      val executable: Array[Int] => Array[Int] = ops(op)
+      executable(input.toArray).toVector
+
+
+
+  private val input: (Vector[Operation], Int) =
+
+    def parseOps(s: String): Option[Operation] =
+      if s.contains("#ip") then None
+      else
+        val in: Vector[String] = s.split(" ").toVector
+        Some(Operation(in(0), in(1).toInt, in(2).toInt, in(3).toInt))
+
+    def parseIp(s: String): Option[Int] =
+      if s.contains("#ip") then
+        Some(s.split(" ")(1).toInt)
+      else None
+
+    val lines: Vector[String] = Source
+      .fromResource(s"day${day}.txt")
+      .getLines
+      .toVector
+
+    (lines.flatMap(parseOps), lines.flatMap(parseIp).head)
+
+  def incrementIp(reg: Vector[Int], ip: Int): Vector[Int] =
+    val tmp: Array[Int] = reg.toArray
+    tmp(ip) = tmp(ip) + 1
+    tmp.toVector
+
+
+  def runProgram(prg: Vector[Operation], ip: Int, n: Int = 0)(maxIts: Int, reg: Vector[Int]): Vector[Int] =
+    if n >= maxIts && maxIts != -1 then reg
+    else
+      val instruction: Try[Operation] = Try(prg(reg(ip)))
+      instruction match
+        case Failure(_) => {println("HIT") ; reg}
```

Comment:
it's not clear to me under which condition the failure can occur :|

Reply:
Hi I've updated the code and provided lots of documentation. Hope that it's a lot clearer now. If not, let me know, then I'll take another look.

I value the feedback a lot btw, so big thanks!!

Reply:
nice koen - very clear writeup :)